### PR TITLE
class library and ide: refactor server

### PR DESCRIPTION
**Pull request by telephon**
Here is the last version of the server refactoring.

### CODE REVIEW

**Comment by telephon** [comment]
I don't understand why this failed. The error is `[31mError:[0m python-2.7.12_1 already installed`. Nothing to do with this PR, really.

**Comment by bagong** [comment]
Retriggering the build "fixed" it ;)

**Comment by telephon** [comment]
excellent.

**Comment by snappizz** [comment]
What should I look for testing this PR?

File: HelpSource/Classes/Recorder.schelp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+ClassMethods::
+
+method::new
+Create a new instance for a given server.
+
+argument::server
+
+
+InstanceMethods::
```
Comment (snappizz) [inline HelpSource/Classes/Recorder.schelp]: Two small doc warnings:

    WARNING: SCDoc: In /home/nathan/git/supercollider/HelpSource/Classes/Recorder.schelp
      Method -prepareForRecord has 2 args, but doc has 1 argument:: tags.
    WARNING: SCDoc: In /home/nathan/git/supercollider/HelpSource/Classes/Recorder.schelp
      Method -record has 5 args, but doc has 1 argument:: tags.

**Comment by telephon** [comment]
@snappizz: A lot of the server class functionality is refactored in this PR, but especially booting, quitting, recording, scoping, pause and resume recording, both programmatically and from the IDE menus would be possible places to look at.

In terms of code quality, the C++/Qt  code may need more eyes perhaps.

File: HelpSource/Classes/Recorder.schelp (hunk 1 above)
Comment (telephon) [inline HelpSource/Classes/Recorder.schelp]: ok, done

**Comment by snappizz** [comment]
Oops, sorry -- approval was by accident

**Comment by snappizz** [comment]
Found a bug! Steps to reproduce:

- boot the server
- kill the scsynth process externally (e.g. `killall scsynth`)
- reboot server

```
ERROR: Message 'disable' not understood.
RECEIVER:
   nil
ARGS:

PROTECTED CALL STACK:
    Meta_MethodError:new    0x1d5c580
        arg this = DoesNotUnderstandError
        arg what = nil
        arg receiver = nil
    Meta_DoesNotUnderstandError:new 0x1d5e540
        arg this = DoesNotUnderstandError
        arg receiver = nil
        arg selector = disable
        arg args = [  ]
    Object:doesNotUnderstand    0x1269540
        arg this = nil
        arg selector = disable
        arg args = nil
    a FunctionDef   0x336fb00
        sourceCode = "<an open Function>"
    a FunctionDef   0x38339c0
        sourceCode = "<an open Function>"
    Function:prTry  0x206f980
        arg this = a Function
        var result = nil
        var thread = a Thread
        var next = nil
        var wasInProtectedFunc = false
    
CALL STACK:
    DoesNotUnderstandError:reportError   0x62f8038
        arg this = <instance of DoesNotUnderstandError>
    < closed FunctionDef >   0x62f7ab8
        arg error = <instance of DoesNotUnderstandError>
    Integer:forBy   0x5aebc48
        arg this = 0
        arg endval = 0
        arg stepval = 2
        arg function = <instance of Function>
        var i = 0
        var j = 0
    SequenceableCollection:pairsDo   0x62fa1e8
        arg this = [*2]
        arg function = <instance of Function>
    Scheduler:seconds_   0x62fa298
        arg this = <instance of Scheduler>
        arg newSeconds = 26.313241528
    Meta_AppClock:tick   0x62fa3f8
        arg this = <instance of Meta_AppClock>
        var saveClock = <instance of Meta_SystemClock>
    Process:tick   0x62fa4a8
        arg this = <instance of Main>
^^ The preceding error dump is for ERROR: Message 'disable' not understood.
RECEIVER: nil
```

**Comment by telephon** [comment]
great! will fix...

**Comment by telephon** [comment]
@snappizz ok, can you pull and test again? It is a good idea to try to do all kinds of things with a now officially unresponsive server, because that is a new "feature".

**Comment by snappizz** [comment]
Thanks, the error is fixed. The behavior of the server status bar is still a little confusing. After the process is killed, the CPU meter hangs for a while and the numbers turn yellow after a few seconds. There isn't notification to the user that the server died, aside from `RESULT = 0` being printed to the post window.

However, the current behavior no better, so I don't consider this a requirement for merge. Up to you if you want to fix it now or later.

**Comment by telephon** [comment]
Yes, it would be much nicer if we would actually know what happened to the server. The thing is that it may well just be hanging for a while, an essentially unknown time, so it would be wrong to just assume it had died. 

Also posting something wouldn't be good, because also when you put your computer to sleep, or load a large sound file, this happens, and you'd see all those strange "server has become unresponsive" messages.

The `RESULT = 0` is interesting. I wonder where sclang gets it from. This may give a clue of how to improve the display in the case when the server was really shot down.

**Comment by snappizz** [comment]
`RESULT = 0` comes from the call to `String:unixCmd` in `Server:bootServerApp`. We can turn it off by setting `postOutput` to false, and we can add a hook using the `action` argument.

**Comment by telephon** [comment]
That's what I had thought as well, but when I call `killall scsynth` from the terminal, I also see it.

**Comment by jamshark70** [comment]
> We can turn it off by setting postOutput to false

It would be worth testing to be sure this doesn't suppress error posting from scsynth. I believe `postOutput: false` will block all stdout, and that would be bad.

**Comment by telephon** [comment]
For what this PR is concerned, I'm rather curious to know where the posting comes from. It could be used at least for local servers to detect a killed server.

**Comment by snappizz** [comment]
I have filed #2452 concerning the inability to disable `RESULT = ` posting without disabling post output entirely.

You can detect a killed server using the `action` argument to unixCmd.

**Comment by telephon** [comment]
>You can detect a killed server using the action argument to unixCmd.

Yes, but only if you have called the command from sclang. Make the following experiment:

Open the terminal (bash or whatever) and type `killall scsynth`. Then look in the post window of sclang. You'll (probably) see `RESULT = 0`. Why is that!?

**Comment by snappizz** [comment]
I think we keep talking past each other here... never mind.

Thanks for your work on this, I'm excited to get this merged. Let's get at least one thorough code review in before merging. Sound good?

I can go over it on Saturday, and of course everyone else should feel free to step up.

**Comment by telephon** [comment]
sounds good!

**Comment by telephon** [comment]
Ah, now I got it: because sclang actually creates the process using a `unixCmd`, we can also test for an end.

see #2458.

**Comment by telephon** [comment]
ping …

**Comment by snappizz** [comment]
sorry i didn't get around to code review as promised :( pretty busy for the next few days unfortunately

**Comment by telephon** [comment]
no problem, just wanted to keep this awake :)

**Comment by telephon** [comment]
ping

**Comment by snappizz** [comment]
I swear I will get around to code review for this at some point, but right now I'm behind on work and school.

**Comment by telephon** [comment]
@snappizz no Problem, this is just for keeping it alive. if others can test too, it wouldn't hurt

**Comment by adcxyz** [comment]
I have browsed through it, and it looks really good. 
Will test for a few more days - it feels very complete and mergeable already. 

best adc

**Comment by jamshark70** [comment]
#2521. It was only a matter of time before we got caught with pants around ankles.

**Comment by telephon** [comment]
that is a rather drastic metaphor for code review, james!

**Comment by jamshark70** [comment]
Oh, heh heh, it's a modified version of the American idiom "(to) get caught with your pants down," meaning to be unprepared for something that you knew was coming up and should have been ready to do. It seemed apt (if colorful).

**Pull request merged**